graal/src/main/java/io/micronaut/graal/reflect/GraalTypeElementVisitor.java

```diff
@@ -216,85 +225,175 @@ public final void finish(VisitorContext visitorContext) {
 
             executed = true;
 
-            List<Map> json;
-            ObjectMapper mapper = new ObjectMapper();
-            ObjectWriter writer = mapper.writer(new DefaultPrettyPrinter());
+            generateNativeImageProperties(visitorContext);
+            generateResourceConfig(visitorContext);
+        }
+    }
+
+    private void generateNativeImageProperties(VisitorContext visitorContext) {
+        List<Map> json;
+        ObjectMapper mapper = new ObjectMapper();
+        ObjectWriter writer = mapper.writer(new DefaultPrettyPrinter());
+
+        Optional<Path> projectDir = visitorContext.projectDir();
+
+        File userReflectJsonFile = projectDir
+                .map(projectPath -> Paths.get(projectPath.toString(), BASE_REFLECT_JSON).toFile())
+                .orElse(null);
+
+        if (userReflectJsonFile != null && userReflectJsonFile.exists()) {
+            try {
+                json = mapper.readValue(userReflectJsonFile, new TypeReference<List<Map>>() {
+                });
+            } catch (Throwable e) {
+                visitorContext.fail("Error parsing base reflect.json: " + BASE_REFLECT_JSON, null);
+                return;
+            }
+        } else {
+            json = new ArrayList<>();
+        }
+
+        if (CollectionUtils.isEmpty(classes) && CollectionUtils.isEmpty(arrays) && CollectionUtils.isEmpty(json)) {
+            return;
+        }
+
+        try {
+            String path = buildNativeImagePath();
+            String reflectFile = path + REFLECTION_CONFIG_JSON;
+            String propsFile = path + NATIVE_IMAGE_PROPERTIES;
+
+            visitorContext.visitMetaInfFile(propsFile).ifPresent(gf -> {
+                visitorContext.info("Writing " + NATIVE_IMAGE_PROPERTIES + " file to destination: " + gf.getName());
+                try (PrintWriter w = new PrintWriter(gf.openWriter())) {
+                    w.println("Args = -H:ReflectionConfigurationResources=${.}/reflection-config.json");
```

Comment:
I don't think we need this business. If a file exists call `reflect-config.json` then GraalVM will pick it up automatically

Reply:
I need to revert the change. The code is necessary to tell GraalVM to pickup the generated file. Without it, all the files generated during Micronaut compilation are not picked up by native-image. I've tested with a sample app it failed.